### ISSUE TITLE
[DISCUSSION] Better pre-release workflow

### ISSUE THREAD

**Issue by Xymanek** [X2CommunityCore/X2WOTCCommunityHighlander#560]
Right our workflow is quite simple:

* People submit PRs
* Repeat previous point many times
* One of the maintainers decides to do a release
* Merges all PRs while resolving conflicts
* Publishes an RC (which is used by like no one)
* Waits ~1 week
* Publishes a workshop release
* We get bug reports and do hotfixes (see v1.18.1 and v1.18.2)

Now, I think the drawbacks are quite clear at this point, but lemme state them just in case:

* Difficult merging/release procedure
* Very little pre-release testing. The little that is done is usually done by mod authors/testers **using the submitted hooks**, leaving the "vanilla behaviour should not be touched by default" mandate completely untested (again, see v1.18.1 and v1.18.2)
* Mod authors have a very hard time using their hooks until the release is published. This luxury is generally enjoyed only by big projects (eg. StratOverhaul, LWotC, RPGO, etc) that have the manpower and knowledge to maintain non-trivial operation of custom CHL "release channel" (merging PRs, cooking, building against, etc). The smaller projects are stuck waiting for the release or begging someone on discord to explain/guide how to make a CHL release with their PR
* Anything that lands in a release is forever mandated to not change (BC policy), so what happens if the hook turns out to be non-great in practice (looking at you, `CanAdd[...]`)? Welp, too bad
* As practice shows, if a PR doesn't land in a release it's doomed to be forever stuck in nether, never to be merged (unless the author ~~annoys maintainers~~ complains enough)
* We are scared to do substantial changes in case "it may break something" since we don't have enough testing

Now that we know what we want to fix/improve, let's see what we can do:

* Merge PRs as soon as they are deemed ready. There is no need for them to sit and rot until it needs to be rebased if the "theoretical" part is done
* Have a beta/nightly release **on workshop** that will be updated with new PRs as they are merged. This can be used by literary everyone - mod authors, testers, simple players who want to contribute to the project
* ???
* Profit

cc @X2CommunityCore/core-highlander-team @pledbrook

**Comment by MrNiceUK** [comment]
Does GitHub provide access to VMs to handle compilation, like Azure (previously simply called VisualStudio) repo hosting does?

**Comment by abatewongc** [comment]
@MrNiceUK 
Yes, via Github Actions. However, it doesn't support Windows and Windows support isn't planned as of last week. Automation for this will probably be tricky.

**Comment by MrNiceUK** [comment]
Hmmm, and porting to Azure isn't an option eh? 😜

**Comment by Xymanek** [comment]
There is https://www.appveyor.com/

**Comment by MrNiceUK** [comment]
What about VM within a VM inceptioning?

**Comment by Xymanek** [comment]
The real issue is that to cook the CHL you need the 90+GB SDK and a steam client running to upload

**Comment by MrNiceUK** [comment]
does using the alternate uploaded get around the steam client requirement?
Just clicked your link @Xymanek, and yeah:
`Support for GitHub, GitHub Enterprise, Bitbucket, GitLab, Azure Repos, Kiln or custom repos`
so if they give you enough VM storage it's an option

**Comment by Musashi1584** [comment]
Here is how i think it could be done in practice:

Every pull request, once marked as ready, will be merged on a to be named branch (e.g. `beta`).

This branch can be fetched by the contributors to test integration of other pr's and a the latest state etc.

When a critical mass of pr's is accumulated and a maintainer has time the beta branch will be cooked and uploaded to the "beta" steam workshop page.
After some time and feedback, regressions and changes can be made in form of new pr's to the beta branch and rolled out to steam again.
Once everyone has the feeling the beta is stable enough it will be merged to the master branch and published on the stable highlander mod page.

This means we need to maintain two different `PublishedFileId.ID` files (one in the beta and one in the master branch) that must not be merged. Git doesn't really provide a way to freeze a file so that it cannot be modified. We need to find a way to handle this. 
You can exclude files from merge but it needs some command line trickery:

```
git merge --no-ff --no-commit beta
git reset HEAD PublishedFileId.ID
git checkout -- PublishedFileId.ID
git commit -m "merged beta"
```

Maybe someone has a better idea here. Thoughts?

**Comment by abatewongc** [comment]
We could also consider using Azure anyways: https://azure.microsoft.com/en-us/services/devops/pipelines/

**Comment by Xymanek** [comment]
@abatewongc looks like they have github integration (which makes sense since MS owns GH) and a nice free plan for OSS projects, but I don't see how much storage they provide for runtime and images/VM templates/whatever you want to call it

**Comment by abatewongc** [comment]
@Xymanek I can't find that information either. I'll try it with RT:TP tonight using the miniSDK container and see if there's any initial issues.

**Comment by MrNiceUK** [comment]
https://docs.microsoft.com/en-gb/azure/devops/pipelines/agents/hosted?view=azure-devops
There's the info on the MS hosted VMs available.

**Comment by MrNiceUK** [comment]
Key point, doesn't look like you can fit the *full* version of the SDK in them...

**Comment by abatewongc** [comment]
Don't see why not, Microsoft might up our disk space if we send in a support ticket.

**Comment by MrNiceUK** [comment]
Well, because the VMs only have just over 10Gb of space to play with (not including tools part of the VM image).

**Comment by abatewongc** [comment]
What I'm hoping is that specifying/loading the container in the .pipeline file won't count towards that 10GB. :pray:

**Comment by MrNiceUK** [comment]
Sure, sounds like you've looked a bit more into how to actually setup/use the pipelines to me, was mainly responding to @Xymanek asking about the CM specs.

**Comment by abatewongc** [comment]
I'm currently seeing an average wait time of ~10 minutes to start the job. Considering that the miniSDK is ~1/50th of the size of the fullSDK, I'd say it's not feasible unless we want to host our own agent somewhere and at that point we probably have better options.

Also, apparently, the container I built is choking on the make 😡

**Comment by Xymanek** [comment]
> we want to host our own agent somewhere

Did a quick look around, we are looking at least $15/mo for the 100gb storage + win license. So unless someone has a homelab, it seems that the automation isn't happening -.-

**Comment by Xymanek** [comment]
In theory, would people be interested in crowd-funding the infrastructure for automated builds? I realize that it's a whole other thing to setup (including legal aspect) but may be the only way if we don't want someone to do the merging and uploading manually

**Comment by yoshimo** [comment]
What amount is needed there?

**Comment by Musashi1584** [comment]
I think crowd-funding a automated build server is a good idea. But first we should evaluate if that really works and how much money we would need. @Xymanek @abatewongc 

Beside that there are a few open ends i like to hear opinions about:

- How do we handle the PublishedFileId? My suggestion is to keep the `PublishedFileId.ID`file empty and automated builds or manual releases need to copy the beta or master PublishedFileId into that file on publish from another file. This should minimize the risk of overriding the id in merges or publish to the wrong workshop item.
- When should PR be closed? When merged into beta or when the beta is merged into the master?

**Comment by Xymanek** [comment]
> What amount is needed there?

I hope to be able to make it work at ~20USD/mo

> When should PR be closed? When merged into beta or when the beta is merged into the master?

The question is best answered by asking "what happens to PR if we find out that it needs to be amended while in beta?"

Also, @abatewongc could you please send me your container? I want to experiment with it

**Comment by abatewongc** [comment]
@Xymanek It doesn't work, I'm unsure as to the exact cause but I believe the DirectX9.0c installation is silently failing.
